index/index.go

```diff
@@ -0,0 +1,393 @@
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+package index
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/gob"
+	"errors"
+	"fmt"
+
+	"github.com/gofrs/uuid"
+	"golang.org/x/crypto/sha3"
+
+	"github.com/cybercryptio/d1-lib/crypto"
+	"github.com/cybercryptio/d1-lib/data"
+	"github.com/cybercryptio/d1-lib/id"
+	"github.com/cybercryptio/d1-lib/io"
+	"github.com/cybercryptio/d1-lib/key"
+)
+
+// The length of the master key.
+const MasterKeyLength = 32
+
+// Error returned if a master key with invalid length is used.
+var ErrInvalidMasterKeyLength = fmt.Errorf("invalid key length, accepted key length is %d bytes", MasterKeyLength)
+
+// Error returned if the caller cannot be authenticated by the Identity Provider.
+var ErrNotAuthenticated = errors.New("user not authenticated")
+
+// Error returned if a user tries to access data they are not authorized for.
+var ErrNotAuthorized = errors.New("user not authorized")
+
+type SecureIndex struct {
+	keyProvider key.Provider
+	ioProvider  io.Provider
+	idProvider  id.Provider
+
+	indexKey []byte
+}
+
+// NewSecureIndex creates a SecureIndex which is used to manage keyword/identifier pairs.
+func NewSecureIndex(keyProvider key.Provider, ioProvider io.Provider, idProvider id.Provider) (SecureIndex, error) {
+	keys, err := keyProvider.GetKeys()
+	if err != nil {
+		return SecureIndex{}, err
+	}
+
+	if len(keys.IEK) != MasterKeyLength {
+		return SecureIndex{}, ErrInvalidMasterKeyLength
+	}
+
+	return SecureIndex{
+		keyProvider: keyProvider,
+		ioProvider:  ioProvider,
+		idProvider:  idProvider,
+		indexKey:    keys.IEK,
+	}, nil
+}
+
+// Add adds a keyword/identifier pair to the secure index.
+func (i *SecureIndex) Add(token, keyword, identifier string) error {
+	identity, err := i.idProvider.GetIdentity(token)
+	if err != nil {
+		return ErrNotAuthenticated
+	}
+	if !identity.Scopes.Contains(id.ScopeIndex) {
+		return ErrNotAuthorized
+	}
+
+	tagger, cryptor, err := i.getTaggerAndCryptor(keyword)
+	if err != nil {
+		return err
+	}
+
+	// Compute the current last sealed Identifier with the given keyword, i.e. the sealed
+	// Identifier with the largest value of the counter.
+	lastIdentifier, err := i.lastIdentifier(keyword)
+	if err != nil {
+		return err
+	}
+
+	counter := uint64(0)
+
+	// If the current last sealed Identifier is not an empty sealed Identifier, i.e. the keyword
+	// has already been added to the secure index, update the counter.
+	if lastIdentifier.Identifier != "" {
+		counter = lastIdentifier.NextCounter
+	}
+
+	// Compute new labelUUID and plaintext Identifier, seal it, and send it to the IO Provider.
+	labelUUID, err := computeLabelUUID(counter, tagger)
+	if err != nil {
+		return err
+	}
+
+	plainID := data.Identifier{Identifier: identifier, NextCounter: counter + 1}
+
+	sealedID, err := plainID.Seal(labelUUID, cryptor)
+	if err != nil {
+		return err
+	}
+	if err = i.putSealedIdentifier(labelUUID, &sealedID); err != nil {
```

Comment:
Nitpick, but this could be:
```suggestion
	if err := i.putSealedIdentifier(labelUUID, &sealedID); err != nil {
```